plugin/completion.py

```diff
@@ -1,25 +1,85 @@
 import sublime
 import sublime_plugin
-from .core.completion import parse_completion_response, format_completion
-from .core.configurations import is_supported_syntax
-from .core.documents import get_document_position, position_is_word
-from .core.edit import parse_text_edit
+
+from .core.protocol import Request, Range, InsertTextFormat
+from .core.settings import settings, client_configs
 from .core.logging import debug
-from .core.protocol import Request
+from .core.completion import parse_completion_response, format_completion
 from .core.registry import session_for_view, client_from_session, LSPViewEventListener
+from .core.configurations import is_supported_syntax
+from .core.documents import get_document_position
 from .core.sessions import Session
-from .core.settings import settings, client_configs
+from .core.edit import parse_text_edit
+from .core.views import range_to_region
 from .core.typing import Any, List, Dict, Tuple, Optional, Union
 
 
-class CompletionState(object):
-    IDLE = 0
-    REQUESTING = 1
-    APPLYING = 2
-    CANCELLING = 3
+last_text_command = None
 
 
-last_text_command = None
+class LspSelectCompletionItemCommand(sublime_plugin.TextCommand):
+    def run(self, edit: 'Any', item: 'Dict') -> None:
+        insert_text_format = item.get("insertTextFormat")
+
+        text_edit = item.get('textEdit')
+        if text_edit:
+            new_text = text_edit.get('newText')
+
+            range = Range.from_lsp(text_edit['range'])
+            edit_region = range_to_region(range, self.view)
+
+            if insert_text_format == InsertTextFormat.Snippet:
+                self.view.run_command("insert_snippet", {"contents": new_text})
```

Comment:
I think handling of the edit region is missing here. Imagine this situation: The user types `aaa$bbb`.
The server returns one completion item with range (2, 6) and textEdit `foobar($0)`. So we should end up with `aafoobar()`. This code will end up with `aaa$bbbfoobar()`.

My guess:

1.   Remove the region from the buffer in all cases: `self.view.erase(edit_region)`.
2a. If it's a Snippet, then run `insert_snippet {"content": new_text}`.
2b. If it's a PlainText, then run `insert_snippet {"content": snippet_escape(new_text)}`, where `snippet_escape` replaces all `$` signs with `\$`.

In particular, I'd like to know if a completion item like this one would work: https://github.com/sublimelsp/LSP/blob/2d7bed9bc58eb25fe670e472813cc810bf77852b/tests/test_completion.py#L351-L370

Here you can see that we have a single letter `I` in the buffer, and the completion is something completely different.

Reply:
Also that completion example from metals is supposed to be a snippet I believe: `def foo: Int = ${1:???}\n   def boo: Int = ${2:???}`

Reply:
Also try this situation:

The buffer is `#include <io|>` ( `|` denotes the cursor position)
The completion is `iostream>` with edit range (10, 12), so it should remove `io>` and replace it with `iostream>`.

Reply:
> 2b. If it's a PlainText, then run insert_snippet {"content": snippet_escape(new_text)}, where snippet_escape replaces all $ signs with \$

Is there any reason why you suggested to run insert_snippet for a PlainText format completion item and escape the $ sign?

Why not just use view.insert() or view.replace()? 
After all I would run `insert_snippet` command only for completion items that are  InsertTextFormat.Snippet.

> I think handling of the edit region is missing here.
Yeah, I will add it :)

Reply:
> Also try this situation:
> 
> The buffer is `#include <io|>` ( `|` denotes the cursor position)
> The completion is `iostream>` with edit range (10, 12), so it should remove `io>` and replace it with `iostream>`.

I tested this with the JSON language server, it works as expected :)

![good](https://user-images.githubusercontent.com/22029477/73687756-00908c80-46cb-11ea-9380-97dbac5aa32c.gif)

Reply:
> Is there any reason why you suggested to run insert_snippet for a PlainText format completion item and escape the $ sign?

You're right, just use self.view.replace(...)